rmw_dds_common/include/rmw_dds_common/graph_cache.hpp

```diff
@@ -0,0 +1,501 @@
+struct EntityInfo;
+
+/// Graph cache data structure.
+/**
+ * Manages relationships between participants, nodes and topics.
+ */
+class GraphCache
+{
+  friend
+  std::ostream &
+  operator<<(std::ostream & ostream, const GraphCache & topic_cache);
+
+public:
+  /**
+   * \defgroup dds_discovery_api
+   * Methods used to update the Graph Cache based on DDS discovery.
+   * @{
+   */
+
+  /**
+   * Add a data writer based on discovery.
+   *
+   * \param gid The data writer guid.
+   * \param topic_name
+   * \param type_name
+   * \param participant_gid gid of the participant.
+   * \param qos Quality of service of the writer.
+   * \return `true` when a change took place.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  bool
+  add_writer(
+    const rmw_gid_t & writer_gid,
+    const std::string & topic_name,
+    const std::string & type_name,
+    const rmw_gid_t & participant_gid,
+    const rmw_qos_profile_t & qos);
+
+  /**
+   * Add a data reader based on discovery.
+   *
+   * \param gid The data reader guid
+   * \param topic_name
+   * \param type_name
+   * \param participant_gid gid of the participant.
+   * \param qos Quality of service of the writer.
+   * \return `true` when a change took place.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  bool
+  add_reader(
+    const rmw_gid_t & reader_gid,
+    const std::string & topic_name,
+    const std::string & type_name,
+    const rmw_gid_t & participant_gid,
+    const rmw_qos_profile_t & qos);
+
+  /**
+   * Add a data reader based on discovery.
+   *
+   * \param gid The data reader guid.
+   * \param topic_name
+   * \param type_name
+   * \param participant_gid gid of the participant.
+   * \param qos Quality of service of the writer.
+   * \param is_reader Adds a reader when `true`, if not a writer.
+   * \return `true` when a change took place.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  bool
+  add_entity(
+    const rmw_gid_t & gid,
+    const std::string & topic_name,
+    const std::string & type_name,
+    const rmw_gid_t & participant_gid,
+    const rmw_qos_profile_t & qos,
+    bool is_reader);
+
+  /**
+   * Remove a data writer based on discovery.
+   *
+   * \param gid The data writer guid.
+   * \return `true` when a change took place.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  bool
+  remove_writer(const rmw_gid_t & gid);
+
+  /**
+   * Remove a data reader based on discovery.
+   *
+   * \param gid The data reader guid.
+   * \return `true` when a change took place.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  bool
+  remove_reader(const rmw_gid_t & gid);
+
+  RMW_DDS_COMMON_PUBLIC
+  bool
+  remove_entity(const rmw_gid_t & gid, bool is_reader);
+
+  /**
+   * @}
+   * \defgroup common_api
+   * Methods used to update the Graph Cache.
+   * @{
+   */
+
+  /**
+   * Remove a participant based on discovery.
+   *
+   * \param participant_gid
+   * \return `true` when a change took place.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  bool
+  remove_participant(const rmw_gid_t & participant_gid);
```

Comment:
I'm also using that here: https://github.com/ros2/rmw_fastrtps/blob/332c7bb7956522c7cb78573b7b2451b7354cfcc5/rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp#L80.

That's the way a remote `Participant` is being removed from `GraphCache`.
The last status message a `Context` send is empty, and in that case the `Participant` is also removed from the cache:
https://github.com/ivanpauno/rmw_dds_common/blob/aae5f17585483b58ad3974545798358317d3cdc3/rmw_dds_common/src/graph_cache.cpp#L137

But if a remote `Context` finishes unexpectedly without sending a message, DDS discovery will detect the dropped `Participant` and also remove it from the cache.

Reply:
Ok, but how does that explain the various doxygen groups?

They seem more confusing that helpful to me. Maybe add more documentation to help people (me) understand the purpose of the groups or otherwise I'd say it's better to remove them and just document each function separately...

Reply:
Yes, I wanted to clarify where the method is supposed to be used, so people don't need to check in another rmw implementation. I will try to improve documentation.